java-manta-client-unshaded/src/main/java/com/joyent/manta/client/MantaBucketListingIterator.java

```diff
@@ -0,0 +1,485 @@
+ * automates that process and abstracts out the details of the paging process.</p>
+ *
+ * @author <a href="https://github.com/nairashwin952013">Ashwin A Nair</a>
+ */
+@SuppressWarnings({"Duplicates"})
+public class MantaBucketListingIterator implements Iterator<Map<String, Object>>,
+        AutoCloseable  {
+    /**
+     * Logger instance.
+     */
+    private static final Logger LOG = LoggerFactory.getLogger(MantaBucketListingIterator.class);
+
+    /**
+     * Shared url encoder instance.
+     */
+    private static final Encoder UTF8_URL_ENCODER = new Encoder(StandardCharsets.UTF_8);
+
+    /**
+     * Maximum number of results to return for a bucket listing.
+     */
+    static final int MAX_RESULTS = 1024;
+
+    /**
+     * Size of result set requested against the Manta API (2-1024).
+     */
+    private final int pagingSize;
+
+    /**
+     * Prefix filter that helps in optimizing a buckets listing.
+     */
+    private volatile String prefix = null;
+
+    /**
+     * Filter to group names with a common prefix ending in its first occurrence.
+     */
+    private Character delimiter = null;
+
+    /**
+     * Path to bucket in which we will iterate through its contents.
+     */
+    private final String path;
+
+    /**
+     * HTTP request helper class.
+     */
+    private final HttpHelper httpHelper;
+
+    /**
+     * The total number of lines that we have iterated through.
+     */
+    private final AtomicLong lines = new AtomicLong(0);
+
+    /**
+     * The next line of data that we haven't iterated to yet.
+     */
+    private final AtomicReference<String> nextLine = new AtomicReference<>();
+
+    /**
+     * Flag indicated if we have finished and there is nothing left to iterate.
+     */
+    private final AtomicBoolean finished = new AtomicBoolean(false);
+
+    /**
+     * Jackson JSON parsing instance.
+     */
+    private final ObjectMapper mapper = MantaObjectMapper.INSTANCE;
+
+    /**
+     * The marker we use to request against the Manta API.
+     */
+    private volatile String marker;
+
+    /**
+     * The current {@link BufferedReader} instance that wraps the HTTP response
+     * {@link java.io.InputStream} from our most recent request to the API.
+     */
+    private volatile BufferedReader br;
+
+    /**
+     * The most recent response object from the page of data that we are currently
+     * parsing.
+     */
+    private volatile CloseableHttpResponse currentResponse;
+
+    /**
+     * Create a new instance of a bucket list iterator.
+     *
+     * @param path path to bucket in which we will iterate through its contents
+     * @param httpHelper HTTP request helper class
+     * @param prefix prefix filter that helps in optimizing a buckets listing
+     * @param pagingSize size of result set requested against the Manta API (2-1024)
+     * @param delimiter filter to group names with a common prefix ending in its first occurrence
+     */
```

Comment:
Don't we also need a ```MantaBucketListingIterator``` constructor that can accept a ```marker``` ?

Reply:
Noted, one more constructor encapsulating `marker` as a parameter will be added.